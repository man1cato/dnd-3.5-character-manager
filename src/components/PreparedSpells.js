import React, {Fragment} from 'react'
import {connect} from 'react-redux'
import update from 'immutability-helper'

import SpellModal from './SpellModal'


const listPreparedSpells = (spellbook) => {
   let preparedSpells = []
   spellbook.forEach((page) => {
      const filteredSpells = page.spells.filter((spell) => spell.prepared > 0)
      filteredSpells.forEach((spell) => preparedSpells.push(spell))         
   })
   return preparedSpells
}

export class PreparedSpells extends React.Component{
   state = {
      preparedSpells: listPreparedSpells(this.props.spellbook),
      spellbook: this.props.spellbook,
      castSpells: [],
      selected: undefined
   }     

   handleOpenModal = (e) => {
      const id = e.target.id;
      const selected = this.props.spells[id]
      this.setState({selected})
   }

   handleCloseModal = () => {
      this.setState({selected: undefined})
   }

   handleChange = (e) => {
      const level = e.target.getAttribute("level")
      const spellId = e.target.getAttribute("spellid")
      const index = this.props.spellbook[level].spells.findIndex((spell) => spell.id === spellId)
      const attribute = e.target.getAttribute("attribute")
      const valueChange = Number(e.target.getAttribute("change"))
      
      this.setState((prevState) => {
         const spell = prevState.spellbook[level].spells[index]
         const value = prevState.spellbook[level].spells[index][attribute] + valueChange
         const remaining = attribute === "prepared" ? value - spell.used : spell.prepared - value
         let castSpells = prevState.castSpells
         if (valueChange > 0) {
            castSpells.push(spellId)
         } else {
            castSpells = castSpells.filter((id) => id !== spellId)
         }

         return {
            spellbook: update(prevState.spellbook, {
               [level]: {
                  spells: {
                     [index]: {
                        [attribute]: { $set: value },
                        remaining: { $set: remaining }
                     }
                  }
               }
            }),
            castSpells
         }
      }, () => {
         this.setState((prevState) => {
            const total = prevState.spellbook[level].spells.map((spell) => spell.prepared).reduce((total, num) => total + num)
            const preparedSpells = listPreparedSpells(prevState.spellbook)
            return {
               spellbook: update(prevState.spellbook, {
                  [level]: {
                     total: { $set: total }
                  }
               }),
               preparedSpells               
            }
         }, () => {
            this.props.startEditProfile(this.props.id, {spellbook: this.state.spellbook})
         })
      })
   }
  
   render () { 
      return (
         <Fragment>
            <SpellModal 
               selected={this.state.selected}
               handleCloseModal={this.handleCloseModal}
            />            

            <h3 className="row row--center">Prepared Spells</h3>

            <div className="grid grid--preparedSpells">
               <h5 className="grid__col1">Spell</h5>
               <h5>Rmng</h5>
               <div></div>
               <div></div>

               {this.state.preparedSpells.map((spell, i) => (
                  <Fragment key={i}>						
                     <button 
                        className="grid__col1 button--link"                         
                        key={`spell${spell.level}${i}`}
                        id={spell.id}
                        onClick={this.handleOpenModal}
                     >
                        {spell.name}
                     </button>
            
                     <div key={`rmng${spell.level}${i}`}>{spell.remaining}</div>

                     {this.state.castSpells.includes(spell.id) ?
                        <button 
                           key={`undo${spell.level}${i}`}
                           spellid={spell.id}
                           level={spell.level}
                           attribute="used"
                           change={-1}
                           onClick={this.handleChange}
                        >
                           Undo
                        </button>
                        :
                        <div></div>
                     }
                     
                     {spell.remaining > 0 ?
                        <button 
                           key={`cast${spell.level}${i}`}
                           spellid={spell.id}
                           level={spell.level}
                           attribute="used"
                           change={1}
                           onClick={this.handleChange}
                        >
                           Cast
                        </button>
                        :
                        <div></div>
                     }

                     
                  </Fragment>
               ))}
            </div>
            
         </Fragment>
      )
   }
}


export default PreparedSpells