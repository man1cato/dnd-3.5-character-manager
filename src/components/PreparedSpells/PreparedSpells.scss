@import '../../styles/base/settings';


.PreparedSpells__grid {
   display: grid;
   grid-template-columns: 3fr repeat(3, 1fr);
   grid-auto-flow: row;
   grid-gap: $s-size;
   justify-items: center;
   align-items: end;
   text-align: center; 
   > input {
      max-width: 50px;
      text-align: center;
   }
}