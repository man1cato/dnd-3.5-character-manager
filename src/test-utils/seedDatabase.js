import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}