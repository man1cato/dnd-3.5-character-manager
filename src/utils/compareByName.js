export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}