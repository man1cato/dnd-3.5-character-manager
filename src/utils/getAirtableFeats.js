import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let feats = []

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const featGroup = res.data.records.map((feat) =>  ({
				id: feat.id,
				name: feat.fields.Name,
				type: feat.fields.Type.join(", ") || null,
				description: feat.fields.Description || null
			}));
			
			feats = feats.concat(featGroup)
		} while (!!offset)
		
		return _.orderBy(feats, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}