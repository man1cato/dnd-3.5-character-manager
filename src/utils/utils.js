import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])