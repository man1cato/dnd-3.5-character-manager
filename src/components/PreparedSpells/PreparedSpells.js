import React, { Fragment, useState, useEffect } from 'react'
import update from 'immutability-helper'

import SpellModal from '../Modals/SpellModal'
import './PreparedSpells.scss'


const listPreparedSpells = (spellbook) => {
	let preparedSpells = []
	spellbook.forEach((page) => {
		const filteredSpells = page.filter((spell) => spell.prepared > 0)
		filteredSpells.forEach((spell) => preparedSpells.push(spell))         
	})
	return preparedSpells
}

export const PreparedSpells = (props) => {
	const [spellbook, setSpellbook] = useState(props.spellbook)
	const [preparedSpells, setPreparedSpells] = useState(listPreparedSpells(spellbook))
	const [castSpells, setCastSpells] = useState([])
	const [clickedSpell, setClickedSpell] = useState(null)

	useEffect(() => {
		setPreparedSpells(listPreparedSpells(spellbook))
		props.handleUpdate({ spellbook })
	}, [spellbook])

	const handleChange = (e) => {
		const level = e.target.getAttribute("level")
		const spellId = e.target.getAttribute("spellid")
		const index = props.spellbook[level].findIndex((spell) => spell.id === spellId)
		const valueChange = Number(e.target.getAttribute("change"))
		
		const spell = spellbook[level][index]
		const used = spell.used + valueChange
		const remaining = spell.prepared - used
		
		if (valueChange > 0) {
			setCastSpells([...castSpells, spellId])
		} else {
			setCastSpells(castSpells.filter((id) => id !== spellId))
		}

		setSpellbook(update(spellbook, {
			[level]: {
				[index]: {
					used: { $set: used },
					remaining: { $set: remaining }
				}			
			}
		}))
	}
  
	return (
		<>
			<div className="PreparedSpells__grid">
				<h5 className="grid__col1">Spell</h5>
				<h5 className="grid__col2">Rmng</h5>

				{preparedSpells.map((spell, i) => (
					<Fragment key={i}>						
						<button 
							className="grid__col1 button--link"                         
							data-testid={spell.id}
							type='button'
							onClick={() => setClickedSpell(props.spells[spell.id])}
						>
							{props.spells[spell.id].name}
						</button>
			
						<div 
							className="grid__col2" 
							data-testid={`${spell.id}Remaining`}
						>
							{spell.remaining}
						</div>

						{castSpells.includes(spell.id) && (
							<button 
								className="grid__col3"
								data-testid={`${spell.id}UndoButton`}
								spellid={spell.id}
								level={spell.level}
								change={-1}
								onClick={(e) => handleChange(e)}
							>
								Undo
							</button>
						)}
						
						{spell.remaining > 0 && (
							<button 
								className="grid__col4"
								data-testid={`${spell.id}CastButton`}
								spellid={spell.id}
								level={spell.level}
								change={1}
								onClick={(e) => handleChange(e)}
							>
								Cast
							</button>
						)}                     
					</Fragment>
				))}
			</div>

			<SpellModal 
				clickedSpell={clickedSpell}
				handleCloseModal={() => setClickedSpell(null)}
			/>   
		</>
	)
}


export default PreparedSpells