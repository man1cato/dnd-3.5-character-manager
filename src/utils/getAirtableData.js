import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}