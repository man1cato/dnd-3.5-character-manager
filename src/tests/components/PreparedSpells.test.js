import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { mount } from 'enzyme'
import _ from 'lodash'

import PreparedSpells from '../../components/PreparedSpells'
import { characterOne } from '../utils/seedDatabase'
import { apiData } from '../utils/utils'


const handleUpdate = jest.fn()
const spellbook = characterOne.spellbook
let props
beforeAll(async () => {
	const api = await apiData()
    props = {
        id: 'characterOne',
        spellbook,
        spells: api.spells,
        handleUpdate
    }
})


test('should render PreparedSpells with profile data', () => {
    const { container } = render(<PreparedSpells {...props} />)
    expect(container.firstChild).toMatchSnapshot()
})

test('should launch SpellModal on feat button click', async () => {
    const { getByText, queryByLabelText, findByLabelText } = render(<PreparedSpells {...props} />)
    const labelMatch = 'Selected Spell'

    expect(queryByLabelText(labelMatch)).toBe(null)

    const textMatch = props.spells[spellbook[0].spells[0].id].name
    fireEvent.click(getByText(textMatch))

    const modal = await findByLabelText(labelMatch)
    expect(modal).toBeDefined()
})

test('should reduce remaining value for spell when Cast button clicked', () => {
    const wrapper = mount(<PreparedSpells {...props} />)
    const { id, remaining } = characterOne.spellbook[0].spells[0]

    expect(wrapper.find(`#${id}Remaining`).text()).toEqual(`${remaining}`)
    expect(wrapper.find(`#${id}UndoButton`)).toHaveLength(0)

    wrapper.find(`#${id}CastButton`).simulate('click')

    expect(wrapper.find(`#${id}Remaining`).text()).toEqual(`${remaining - 1}`)
    expect(wrapper.find(`#${id}UndoButton`)).toHaveLength(1)
})