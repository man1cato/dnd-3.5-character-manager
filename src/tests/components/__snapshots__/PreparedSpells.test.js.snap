// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PreparedSpells with profile data 1`] = `
<div
  class="grid--spells"
>
  <h5
    class="grid__col1"
  >
    Spell
  </h5>
  <h5
    class="grid__col2"
  >
    Rmng
  </h5>
  <button
    class="grid__col1 button--link"
    id="rec8IQeYmBr3f1BMC"
    type="button"
  >
    Acid Splash
  </button>
  <div
    class="grid__col2"
    id="rec8IQeYmBr3f1BMCRemaining"
  >
    1
  </div>
  <button
    change="1"
    class="grid__col4"
    id="rec8IQeYmBr3f1BMCCastButton"
    level="0"
    spellid="rec8IQeYmBr3f1BMC"
  >
    Cast
  </button>
  <button
    class="grid__col1 button--link"
    id="recwpeRoCsea4vidW"
    type="button"
  >
    Arcane Mark
  </button>
  <div
    class="grid__col2"
    id="recwpeRoCsea4vidWRemaining"
  >
    1
  </div>
  <button
    change="1"
    class="grid__col4"
    id="recwpeRoCsea4vidWCastButton"
    level="0"
    spellid="recwpeRoCsea4vidW"
  >
    Cast
  </button>
  <button
    class="grid__col1 button--link"
    id="recRhVs0ahtQZpbXx"
    type="button"
  >
    Charm Person
  </button>
  <div
    class="grid__col2"
    id="recRhVs0ahtQZpbXxRemaining"
  >
    0
  </div>
</div>
`;
